warp_drive/training/models/fully_connected.py

```diff
@@ -69,53 +98,139 @@ def __init__(self, env, model_config, policy, policy_tag_to_agent_id_map):
+        else:
+            raise ValueError(
+                "num_agents can only be the first "
+                "or the last dimension in the observations."
+            )
+        return obs.reshape(num_envs, num_agents, -1)
+
+    def get_flattened_obs(self):
+        """
+        If the obs is of Box type, it will already be flattened.
+        If the obs is of Dict type, then concatenate all the
+        obs values and flatten them out.
+        Returns the concatenated and flattened obs.
+
+        """
+        if isinstance(self.observation_space, Box):
+            if self.create_separate_placeholders_for_each_policy:
+                obs = self.env.cuda_data_manager.data_on_device_via_torch(
+                    f"{_OBSERVATIONS}_{self.policy}"
                 )
-            if not self.fast_forward_mode:
+            else:
+                obs = self.env.cuda_data_manager.data_on_device_via_torch(_OBSERVATIONS)
+
+            flattened_obs = self.reshape_and_flatten_obs(obs)
+        elif isinstance(self.observation_space, Dict):
+            obs_dict = {}
+            for key in self.observation_space:
+                if self.create_separate_placeholders_for_each_policy:
+                    obs = self.env.cuda_data_manager.data_on_device_via_torch(
+                        f"{_OBSERVATIONS}_{self.policy}_{key}"
+                    )
+                else:
+                    obs = self.env.cuda_data_manager.data_on_device_via_torch(
+                        f"{_OBSERVATIONS}_{key}"
+                    )
+
+                if key == _ACTION_MASK:
+                    self.action_mask = self.reshape_and_flatten_obs(obs)
+                else:
+                    obs_dict[key] = obs
+
+            flattened_obs_dict = {}
+            for key in obs_dict:
+                flattened_obs_dict[key] = self.reshape_and_flatten_obs(obs_dict[key])
+            flattened_obs = torch.cat(list(flattened_obs_dict.values()), dim=-1)
+        else:
+            raise NotImplementedError("Observation space must be of Box or Dict type")
+        return flattened_obs
+
+    def forward(self, obs=None, batch_index=None, batch_size=None):
+        """
+        Forward pass through the model.
+        Returns action probabilities and value functions.
+        """
+        if obs is None:
+            assert batch_index < batch_size
+            obs = self.get_flattened_obs()
+
+            if not self.create_separate_placeholders_for_each_policy:
                 agent_ids_for_policy = self.policy_tag_to_agent_id_map[self.policy]
-                self.env.cuda_data_manager.data_on_device_via_torch(name=name)[
-                    batch_index, :, agent_ids_for_policy
-                ] = obs[:, agent_ids_for_policy]
                 ip = obs[:, agent_ids_for_policy]
             else:
-                self.env.cuda_data_manager.data_on_device_via_torch(name=name)[
-                    batch_index
-                ] = obs
                 ip = obs
+
```

Comment:
This is optional though. I feel this forward function serves two purposes, one is forward prediction, the other is feed obs batch, is it better to factorize a function to do the 2nd job here because it might be confusing people

Reply:
Yes, I think it makes a lot of sense.